### ISSUE TITLE
increasing ResNet output size by changing block.expansion does not work

### ISSUE THREAD

**Issue by esarf** [pytorch/vision#5709]
Hello,

I am working with ResNet-18 and actually, after discarding the last self.fc layer, we obtain an output tensor of shape [1,512]. For my research, I would like to obtain different sizes such as [1,1024] or [1,2048] for instances. 

## **What works**

Inside the ResNet, changing the final output channel sizes works perfectly, namely:

```python
self.layer1 = self._make_layer(block, 64, layers[0])
self.layer2 = self._make_layer(block, 128 , layers[1], stride=2, dilate=replace_stride_with_dilation[0])
self.layer3 = self._make_layer(block, 256, layers[2], stride=2, dilate=replace_stride_with_dilation[1])
self.layer4 = self._make_layer(block, 512 * 2, layers[3], stride=2, dilate=replace_stride_with_dilation[2])
```

I just added the * 2 factor and everything behaves as expected. 

## **What does not work**

Reading the code and the paper, it appears that increasing the `block.expansion` parameter from 1 to 2 inside our BottleNeck (or BasicBlock in the ResNet-18 case) should expand the output channel dimension for each layer (see the line `self.inplanes = planes * block.expansion` inside the `_make_layer()` function), but it does not work and gives the following error at the first forward pass. My input tensor is of size [1,1,512,512]. I changed the input channel of the first Conv2d accordingly.  

```python
[<ipython-input-61-1b2c58243ae8>](https://vzt71jrqtl-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220329-060048-RC00_437987085#) in forward(self, x)
    103         print(identity.shape)
    104 
--> 105         out += identity
    106         out = self.relu(out)
    107 

RuntimeError: The size of tensor a (64) must match the size of tensor b (128) at non-singleton dimension 1
```

## **Possible explanation**

The `forward` function of the `BasicBlock` module is the following: 

```python
def forward(self, x: Tensor) -> Tensor:
      identity = x

      print(identity.shape)

      # torch.Size([1, 64, 128, 128])     
      # torch.Size([1, 64, 128, 128])

      out = self.conv1(x)
      out = self.bn1(out)
      out = self.relu(out)

      out = self.conv2(out)
      out = self.bn2(out)

      print(out.shape)

      print(self.downsample)

      if self.downsample is not None:
          identity = self.downsample(x)

      print(identity.shape)

      out += identity
      out = self.relu(out)

      return out
```

with the prints I've added, we can see the shapes before the error happens and the downsample attribute:

```python
identity: torch.Size([1, 64, 128, 128])
out: torch.Size([1, 64, 128, 128])
downsample: Sequential(
  (0): Conv2d(64, 128, kernel_size=(1, 1), stride=(1, 1), bias=False)
  (1): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
)
identity (after downsample): torch.Size([1, 128, 128, 128])
```
 In other words, it looks like the `downsample` increases the size of `identity` but it does not match the shape of the `out` tensor. So I would like to know if I missed something I had to add, in addition to the `block.expansion` parameter ? I tried to increase the `in_planes` and `planes` parameters without success...

I might not be the first who wants to tune this expansion parameter, and even if I already have a solution that answers to my problem, I would like to know where I miss the point.

Thanks a lot!

**Comment by datumbox** [comment]
@esarf Thanks for reporting this.

Could you please provide a snippet that fully reproduces this and the patch/diff of what you try to modify on the TorchVision code? This will help us investigate.

**Comment by esarf** [comment]
@datumbox Thanks for your answer. Here is the full code with my modifications: 

```python
def conv3x3(in_planes: int, out_planes: int, stride: int = 1, groups: int = 1, dilation: int = 1) -> nn.Conv2d:
    """3x3 convolution with padding"""
    return nn.Conv2d(
        in_planes,
        out_planes,
        kernel_size=3,
        stride=stride,
        padding=dilation,
        groups=groups,
        bias=False,
        dilation=dilation,
    )


def conv1x1(in_planes: int, out_planes: int, stride: int = 1) -> nn.Conv2d:
    """1x1 convolution"""
    return nn.Conv2d(in_planes, out_planes, kernel_size=1, stride=stride, bias=False)


class BasicBlock(nn.Module):
    expansion: int = 2      # MODIF: put 2 instead of 1 here 

    def __init__(
        self,
        inplanes: int,
        planes: int,
        stride: int = 1,
        downsample: Optional[nn.Module] = None,
        groups: int = 1,
        base_width: int = 64,
        dilation: int = 1,
        norm_layer: Optional[Callable[..., nn.Module]] = None,
    ) -> None:
        super().__init__()
        if norm_layer is None:
            norm_layer = nn.BatchNorm2d
        if groups != 1 or base_width != 64:
            raise ValueError("BasicBlock only supports groups=1 and base_width=64")
        if dilation > 1:
            raise NotImplementedError("Dilation > 1 not supported in BasicBlock")
        # Both self.conv1 and self.downsample layers downsample the input when stride != 1
        self.conv1 = conv3x3(inplanes, planes, stride)
        self.bn1 = norm_layer(planes)
        self.relu = nn.ReLU(inplace=True)
        self.conv2 = conv3x3(planes, planes)
        self.bn2 = norm_layer(planes)
        self.downsample = downsample
        self.stride = stride

    def forward(self, x: Tensor) -> Tensor:
        identity = x

        print(identity.shape)

        out = self.conv1(x)
        out = self.bn1(out)
        out = self.relu(out)

        out = self.conv2(out)
        out = self.bn2(out)

        print(out.shape)
       
        print(self.downsample)

        if self.downsample is not None:
            identity = self.downsample(x)

        print(identity.shape)

        out += identity
        out = self.relu(out)

        return out


class Bottleneck(nn.Module):
    # Bottleneck in torchvision places the stride for downsampling at 3x3 convolution(self.conv2)
    # while original implementation places the stride at the first 1x1 convolution(self.conv1)
    # according to "Deep residual learning for image recognition"https://arxiv.org/abs/1512.03385.
    # This variant is also known as ResNet V1.5 and improves accuracy according to
    # https://ngc.nvidia.com/catalog/model-scripts/nvidia:resnet_50_v1_5_for_pytorch.

    expansion: int = 4

    def __init__(
        self,
        inplanes: int,
        planes: int,
        stride: int = 1,
        downsample: Optional[nn.Module] = None,
        groups: int = 1,
        base_width: int = 64,
        dilation: int = 1,
        norm_layer: Optional[Callable[..., nn.Module]] = None,
    ) -> None:
        super().__init__()
        if norm_layer is None:
            norm_layer = nn.BatchNorm2d
        width = int(planes * (base_width / 64.0)) * groups
        # Both self.conv2 and self.downsample layers downsample the input when stride != 1
        self.conv1 = conv1x1(inplanes, width)
        self.bn1 = norm_layer(width)
        self.conv2 = conv3x3(width, width, stride, groups, dilation)
        self.bn2 = norm_layer(width)
        self.conv3 = conv1x1(width, planes * self.expansion)
        self.bn3 = norm_layer(planes * self.expansion)
        self.relu = nn.ReLU(inplace=True)
        self.downsample = downsample
        self.stride = stride

    def forward(self, x: Tensor) -> Tensor:
        identity = x

        print(identity.shape)

        out = self.conv1(x)
        out = self.bn1(out)
        out = self.relu(out)

        out = self.conv2(out)
        out = self.bn2(out)
        out = self.relu(out)

        out = self.conv3(out)
        out = self.bn3(out)

        print(out.shape)

        print(self.downsample)

        if self.downsample is not None:
            identity = self.downsample(x)

        out += identity
        out = self.relu(out)

        return out


class ResNet(nn.Module):
    def __init__(
        self,
        block: Type[Union[BasicBlock, Bottleneck]],
        layers: List[int],
        num_classes: int = 1000,
        zero_init_residual: bool = False,
        groups: int = 1,
        width_per_group: int = 64,
        replace_stride_with_dilation: Optional[List[bool]] = None,
        norm_layer: Optional[Callable[..., nn.Module]] = None,
    ) -> None:
        super().__init__()
        _log_api_usage_once(self)
        if norm_layer is None:
            norm_layer = nn.BatchNorm2d
        self._norm_layer = norm_layer

        self.inplanes = 64
        self.dilation = 1
        if replace_stride_with_dilation is None:
            # each element in the tuple indicates if we should replace
            # the 2x2 stride with a dilated convolution instead
            replace_stride_with_dilation = [False, False, False]
        if len(replace_stride_with_dilation) != 3:
            raise ValueError(
                "replace_stride_with_dilation should be None "
                f"or a 3-element tuple, got {replace_stride_with_dilation}"
            )
        self.groups = groups
        self.base_width = width_per_group
        self.conv1 = nn.Conv2d(1, self.inplanes, kernel_size=7, stride=2, padding=3, bias=False)    # MODIF: put 1 channel instead of 3 here
        self.bn1 = norm_layer(self.inplanes)
        self.relu = nn.ReLU(inplace=True)
        self.maxpool = nn.MaxPool2d(kernel_size=3, stride=2, padding=1)
        self.layer1 = self._make_layer(block, 64, layers[0])
        self.layer2 = self._make_layer(block, 128, layers[1], stride=2, dilate=replace_stride_with_dilation[0])
        self.layer3 = self._make_layer(block, 256, layers[2], stride=2, dilate=replace_stride_with_dilation[1])
        self.layer4 = self._make_layer(block, 512, layers[3], stride=2, dilate=replace_stride_with_dilation[2])
        self.avgpool = nn.AdaptiveAvgPool2d((1, 1))
        self.fc = nn.Linear(512 * block.expansion, num_classes)

        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                nn.init.kaiming_normal_(m.weight, mode="fan_out", nonlinearity="relu")
            elif isinstance(m, (nn.BatchNorm2d, nn.GroupNorm)):
                nn.init.constant_(m.weight, 1)
                nn.init.constant_(m.bias, 0)

        # Zero-initialize the last BN in each residual branch,
        # so that the residual branch starts with zeros, and each residual block behaves like an identity.
        # This improves the model by 0.2~0.3% according to https://arxiv.org/abs/1706.02677
        if zero_init_residual:
            for m in self.modules():
                if isinstance(m, Bottleneck):
                    nn.init.constant_(m.bn3.weight, 0)  # type: ignore[arg-type]
                elif isinstance(m, BasicBlock):
                    nn.init.constant_(m.bn2.weight, 0)  # type: ignore[arg-type]

    def _make_layer(
        self,
        block: Type[Union[BasicBlock, Bottleneck]],
        planes: int,
        blocks: int,
        stride: int = 1,
        dilate: bool = False,
    ) -> nn.Sequential:
        norm_layer = self._norm_layer
        downsample = None
        previous_dilation = self.dilation
        if dilate:
            self.dilation *= stride
            stride = 1
        if stride != 1 or self.inplanes != planes * block.expansion:
            downsample = nn.Sequential(
                conv1x1(self.inplanes, planes * block.expansion, stride),
                norm_layer(planes * block.expansion),
            )

        layers = []
        layers.append(
            block(
                self.inplanes, planes, stride, downsample, self.groups, self.base_width, previous_dilation, norm_layer
            )
        )
        self.inplanes = planes * block.expansion
        for _ in range(1, blocks):
            layers.append(
                block(
                    self.inplanes,
                    planes,
                    groups=self.groups,
                    base_width=self.base_width,
                    dilation=self.dilation,
                    norm_layer=norm_layer,
                )
            )

        return nn.Sequential(*layers)

    def _forward_impl(self, x: Tensor) -> Tensor:
        # See note [TorchScript super()]
        x = self.conv1(x)
        x = self.bn1(x)
        x = self.relu(x)
        x = self.maxpool(x)

        x = self.layer1(x)
        x = self.layer2(x)
        x = self.layer3(x)
        x = self.layer4(x)

        x = self.avgpool(x)
        x = torch.flatten(x, 1)
        #x = self.fc(x)

        return x

    def forward(self, x: Tensor) -> Tensor:
        return self._forward_impl(x)


def _resnet(
    block: Type[Union[BasicBlock, Bottleneck]],
    layers: List[int],
    progress: bool,
    **kwargs: Any,
) -> ResNet:
    
    model = ResNet(block, layers, **kwargs)

    return model


def resnet18(*, progress: bool = True, **kwargs: Any) -> ResNet:
    r"""ResNet-18 model from
    `"Deep Residual Learning for Image Recognition" <https://arxiv.org/pdf/1512.03385.pdf>`_.

    Args:
        weights (ResNet18_Weights, optional): The pretrained weights for the model
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    weights = None

    return _resnet(BasicBlock, [2, 2, 2, 2], progress, **kwargs)

model = resnet18() ```

As you can see, I only changed the first convolution (1 channel instead of 3) and the expansion parameter in the BasicBlock with 2 instead of 1. I tested the forward pass with a random image: 

```python
img = torch.randn(1,1,512,512)
model(img)
```

And got this immediately:

```
torch.Size([1, 64, 128, 128])     # identity.shape
torch.Size([1, 64, 128, 128])     # out.shape
Sequential(
  (0): Conv2d(64, 128, kernel_size=(1, 1), stride=(1, 1), bias=False)
  (1): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
)                                               # downsample
torch.Size([1, 128, 128, 128])   # identity.shape after downsample
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
[<ipython-input-3-748d27272d07>](https://localhost:8080/#) in <module>()
      1 img = torch.randn(1,1,512,512)
      2 
----> 3 model(img).shape

6 frames
[<ipython-input-2-9cbb4e23f480>](https://localhost:8080/#) in forward(self, x)
     96         print(identity.shape)
     97 
---> 98         out += identity
     99         out = self.relu(out)
    100 

RuntimeError: The size of tensor a (64) must match the size of tensor b (128) at non-singleton dimension 1
```

As you can see we just have tensors with different shapes between out and identity due to the downsample operation. So I do not really know how to solve it. I must miss something in the deep understanding of ResNet, however I thought it was clear for me.

Thank you!
Emma

**Comment by datumbox** [comment]
@esarf Thanks for the info.

I looked into this and I believe that the `expansion` property is not meant to be overwritten like that. It's a read-only internal property used by the private `_make_layer` method so that it figure out the number of channels. Note that modifying the expansion is also the wrong thing to do here as it will have global effect (it's a class-level property). Moreover the `BasicBlock` is not supposed to expand the input channels or apply the bottleneck logic.

On the other hand, your approach of doubling the output channels when you call `_make_layer`, is the right approach. Of course this is not the approach that the ResNet paper recommends and that's why our implementation doesn't make this configurable. Doubling the channels, while halving the edge of the input if an extremely standard approach to scale networks so adding a property to overwrite this is not something worth doing here.

As far as I see there are 3 options for you:
1. Create a custom ResNet class that inherits from TorchVision's `ResNet` class to make the necessary modifications on the model. Note that doing so, you would be forced to use private methods of the class that might change on the future. An alternative if you want stability is to copy the implementation.
2. Do "model surgery" to replace the old `layer4` block with something that meets your needs. You can do this simply by assigning your own custom block `model.layer4 = custom_conv_block`.
3. Instead of modifying the architecture just add a projection layer from 512 to 1024 channels at the end. This way, you won't have to deal with custom implementations. Whether this makes sense depends on your use-case.

I'm going to close the issue as I think the above answers the question but if you still face problems feel free to reopen.

**Comment by qqqqqqy0227** [comment]
So in my opinion, such trick will work if you use bottleneck to initialize the resnet, but won't work when you use basicblock since basicblock dose not use self.expansion at all?
